### ISSUE TITLE
perl debugger 'n' steps into lvalue subs (not over)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12696]
Migrated from [rt.perl.org#116358](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=116358) (status was 'resolved')

Searchable as RT116358$

**Comment by p5pRT** [comment]
### From devel.chm.01@gmail.com

#### Created by devel.chm.01@gmail.com



Using the perl5 debugger on code with lvalue
subs and stepping via the 'n' command always
steps \_into\_ the lvalue sub rather than over\.
This makes it very difficult to debug lvalue
subs\.  Here is a sample program that can be
run with perl \-d to illustrate the problem&#8203;:

&nbsp;    \#\!/usr/bin/perl
&nbsp;    \#
&nbsp;    \# Illustrate lvalue sub debug problem
&nbsp;    \# The n command steps into the lvalue sub
&nbsp;    \#

&nbsp;    my $data = '';

&nbsp;    sub lslice :lvalue \{
&nbsp;       my \($arg1\, $val1\) = @&#8203;\_;
&nbsp;       print "\.\. in lslice now\\n";
&nbsp;       print "\.\. $arg1 \+ $val1 = "\, $arg1\+$val1\, "\\n";
&nbsp;       $data;
&nbsp;    \}

&nbsp;    print "Start test\, \\$data=$data\\n";
&nbsp;    lslice\(3\,5\) = 4;
&nbsp;    print "Stop  test\, \\$data=$data\\n";

I've marked this report as medium severity but from
the point of users of PDL it makes it very difficult
to debug code and makes things inpenetrable to new
PDL users since if they have a problem and step through
their code they get dumped into some internal PDL subs
that a not relevant to their problem or their stage of
learning\.  Beginner problems are extremely likely to
be their usage and not the core PDL code\.

Thanks much\,
Chris

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium
    module=DB

Site configuration information for perl 5.14.2:

Configured by rurban at Thu Jul 12 13:58:56 CDT 2012.

Summary of my perl5 (revision 5 version 14 subversion 2) configuration:

  Platform:
    osname=cygwin, osvers=1.7.15(0.26053), archname=cygwin-thread-multi-64int
    uname='cygwin_nt-5.1 winxp 1.7.15(0.26053) 2012-05-09 10:25 i686 cygwin '
    config_args='-de -Dlibperl=cygperl5_14.dll -Dcc=gcc-4 -Dld=g++-4
-Darchname=i686-cygwin-threads-64int -Dmksymlinks -Dusethreads
-Accflags=-g'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc-4', ccflags ='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -g
-fno-strict-aliasing -pipe -fstack-protector',
    optimize='-O3',
    cppflags='-DPERL_USE_SAFE_PUTENV -U__STRICT_ANSI__ -g
-fno-strict-aliasing -pipe -fstack-protector'
    ccversion='', gccversion='4.5.3', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8,
Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='g++-4', ldflags =' -Wl,--enable-auto-import
-Wl,--export-all-symbols -Wl,--enable-auto-image-base
-fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /lib
    libs=-lgdbm -ldb -ldl -lcrypt -lgdbm_compat
    perllibs=-ldl -lcrypt
    libc=/usr/lib/libc.a, so=dll, useshrplib=true, libperl=cygperl5_14.dll
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' --shared  -Wl,--enable-auto-import
-Wl,--export-all-symbols -Wl,--enable-auto-image-base -L/usr/local/lib
-fstack-protector'

Locally applied patches:
    Bug#55162 File::Spec::case_tolerant performance
    CYG07 $vendorarch/auto/.rebase
    CYG15 static Win32CORE
    CYG17 cyg-1.7 paths-utf8
    0c612ce82 Fix building static extensions on cygwin, -UUSEIMPORTLIB
    1bac5ecc1 Fix 64-bit threading sv.c: S_anonymise_cv_maybe
    Cygwin::sync_winenv added


@INC for perl 5.14.2:
    /cygdrive/f/perl/local/lib/perl5/cygwin-thread-multi-64int
    /cygdrive/f/perl/local/lib/perl5
    /usr/lib/perl5/site_perl/5.14/i686-cygwin-threads-64int
    /usr/lib/perl5/site_perl/5.14
    /usr/lib/perl5/vendor_perl/5.14/i686-cygwin-threads-64int
    /usr/lib/perl5/vendor_perl/5.14
    /usr/lib/perl5/5.14/i686-cygwin-threads-64int
    /usr/lib/perl5/5.14
    /usr/lib/perl5/site_perl/5.10
    /usr/lib/perl5/vendor_perl/5.10
    /usr/lib/perl5/site_perl/5.8
    .


Environment for perl 5.14.2:
    HOME=/home/chm
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/cygdrive/f/perl/local/bin:/usr/local/bin:/usr/bin:/cygdrive/c/Windows/system32
    PERL5LIB=/cygdrive/f/perl/local/lib/perl5
    PERLDOC_PAGER=less
    PERL_BADLANG (unset)
    PERL_RL=perl
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rocky

I believe this bug is due to the order of these two statements around line
4249 of perl5db\.pl \(5\.17\.7\)

&nbsp;    \# Pop the single\-step value back off the stack\.
&nbsp;    $single |= $stack\[ $stack\_depth\-\- \];

&nbsp;    \# call the original lvalue sub\.
&nbsp;    &$sub;

which should be reversed as it is in \*DB&#8203;::sub\(\)\*\. However it is not enough
just to change these two Perl statements because one needs to save and
return the return value of \*&$sub\(\)\*\. So again one should consult the
additional code that is found in \*DB&#8203;::sub\(\)\* for saving and returning the
return value\. The complication in the code here is figuring out what kind
of value to return\.

\- \- \- \- \-

The rest of this email is a plug for
Devel&#8203;::Trepan\<https://metacpan.org/module/Devel::Trepan>,
so you can stop reading here if you are just interested in this bug for
perl5db\.pl

First I should mention that in order to track down what I think location of
this bug\, it was easier for me to understand what is going on and where to
look by looking at the Devel&#8203;::Trepan source code\. At least for me\, having
smaller more modular files makes it easier for me to narrow my attention
and test things\.  The Package in question here is called DB&#8203;::Sub which is
only 220 or so lines rather than over 10\,000 for perl5db\.pl\.

And the test framework makes it easy for me to work up a test case\.

However I should also mention that Devel&#8203;::Trepan\* also\* had a bug here\, but
for a totally different reason\. And that too is worth mentioning\.

Devel&#8203;::Trepan doesn't use the low\-level "next" bit in set in
\*$DB&#8203;::single\*variable because I have found other bugs in using that
bit\. I don't have
recorded in comments exactly what those bugs are\, but probably one can
browse the git commit history to see when that bit got unused and look at
the integration test that was added at the same time for a more detailed
understanding of what the issue is\. I have a vague recollection that it was
broken in the presence of exception handling\,  handling return events and
thread switching\. It is possible this isn't a problem or as much of a
problem in perl5db\.pl because it doesn't support stopping at returns\,
changing return values or possibly have much support for handling exception
events\.

The disadvantage in Devel&#8203;::Trepan of not using he "next" bit is that
next'ing \("step over"\)  in Devel&#8203;::Trepan can be slower&#8203;: more of this
testing is going on inside Perl code rather than presumably the runtime\.

But at any rate\, with commit
a45169d\<https://github.com/rocky/Perl-Devel-Trepan/commit/a45169d68a3bf49c61c2324e7674e7d1cb86b0e8>this
bug should be fixed in Devel&#8203;::Trepan\. If you are willing to use what's
in git\, you can try it out now\. After the next release\, in contrast to
perl5db\.pl\, prior releases of Perl down to 5\.8\.9 or so will also benefit
from this fix\. This is not an inherent advantage in the Devel&#8203;::Trepan code
so much as the fact that it is a separate CPAN module\. So the same kind of
thing in theory could be done for perl5db\.pl\.


On Fri\, Jan 11\, 2013 at 10&#8203;:58 AM\, Chris Marshall
\<perlbug\-followup@&#8203;perl\.org>wrote&#8203;:

> \# New Ticket Created by  Chris Marshall
> \# Please include the string&#8203;:  \[perl \#116358\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=116358 >
>
>
> This is a bug report for perl from devel\.chm\.01@&#8203;gmail\.com\,
> generated with the help of perlbug 1\.39 running under perl 5\.14\.2\.
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> Using the perl5 debugger on code with lvalue
> subs and stepping via the 'n' command always
> steps \_into\_ the lvalue sub rather than over\.
> This makes it very difficult to debug lvalue
> subs\.  Here is a sample program that can be
> run with perl \-d to illustrate the problem&#8203;:
>
>     \#\!/usr/bin/perl
>     \#
>     \# Illustrate lvalue sub debug problem
>     \# The n command steps into the lvalue sub
>     \#
>
>     my $data = '';
>
>     sub lslice :lvalue \{
>        my \($arg1\, $val1\) = @&#8203;\_;
>        print "\.\. in lslice now\\n";
>        print "\.\. $arg1 \+ $val1 = "\, $arg1\+$val1\, "\\n";
>        $data;
>     \}
>
>     print "Start test\, \\$data=$data\\n";
>     lslice\(3\,5\) = 4;
>     print "Stop  test\, \\$data=$data\\n";
>
> I've marked this report as medium severity but from
> the point of users of PDL it makes it very difficult
> to debug code and makes things inpenetrable to new
> PDL users since if they have a problem and step through
> their code they get dumped into some internal PDL subs
> that a not relevant to their problem or their stage of
> learning\.  Beginner problems are extremely likely to
> be their usage and not the core PDL code\.
>
> Thanks much\,
> Chris
>
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=library
>     severity=medium
>     module=DB
> \-\-\-
> Site configuration information for perl 5\.14\.2&#8203;:
>
> Configured by rurban at Thu Jul 12 13&#8203;:58&#8203;:56 CDT 2012\.
>
> Summary of my perl5 \(revision 5 version 14 subversion 2\) configuration&#8203;:
>
>   Platform&#8203;:
>     osname=cygwin\, osvers=1\.7\.15\(0\.26053\)\,
> archname=cygwin\-thread\-multi\-64int
>     uname='cygwin\_nt\-5\.1 winxp 1\.7\.15\(0\.26053\) 2012\-05\-09 10&#8203;:25 i686
> cygwin '
>     config\_args='\-de \-Dlibperl=cygperl5\_14\.dll \-Dcc=gcc\-4 \-Dld=g\+\+\-4
> \-Darchname=i686\-cygwin\-threads\-64int \-Dmksymlinks \-Dusethreads
> \-Accflags=\-g'
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     useithreads=define\, usemultiplicity=define
>     useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
>     use64bitint=define\, use64bitall=undef\, uselongdouble=undef
>     usemymalloc=n\, bincompat5005=undef
>   Compiler&#8203;:
>     cc='gcc\-4'\, ccflags ='\-DPERL\_USE\_SAFE\_PUTENV \-U\_\_STRICT\_ANSI\_\_ \-g
> \-fno\-strict\-aliasing \-pipe \-fstack\-protector'\,
>     optimize='\-O3'\,
>     cppflags='\-DPERL\_USE\_SAFE\_PUTENV \-U\_\_STRICT\_ANSI\_\_ \-g
> \-fno\-strict\-aliasing \-pipe \-fstack\-protector'
>     ccversion=''\, gccversion='4\.5\.3'\, gccosandvers=''
>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=12345678
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>     ivtype='long long'\, ivsize=8\, nvtype='double'\, nvsize=8\,
> Off\_t='off\_t'\, lseeksize=8
>     alignbytes=8\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='g\+\+\-4'\, ldflags =' \-Wl\,\-\-enable\-auto\-import
> \-Wl\,\-\-export\-all\-symbols \-Wl\,\-\-enable\-auto\-image\-base
> \-fstack\-protector \-L/usr/local/lib'
>     libpth=/usr/local/lib /usr/lib /lib
>     libs=\-lgdbm \-ldb \-ldl \-lcrypt \-lgdbm\_compat
>     perllibs=\-ldl \-lcrypt
>     libc=/usr/lib/libc\.a\, so=dll\, useshrplib=true\, libperl=cygperl5\_14\.dll
>     gnulibc\_version=''
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
>     cccdlflags=' '\, lddlflags=' \-\-shared  \-Wl\,\-\-enable\-auto\-import
> \-Wl\,\-\-export\-all\-symbols \-Wl\,\-\-enable\-auto\-image\-base \-L/usr/local/lib
> \-fstack\-protector'
>
> Locally applied patches&#8203;:
>     Bug\#55162 File&#8203;::Spec&#8203;::case\_tolerant performance
>     CYG07 $vendorarch/auto/\.rebase
>     CYG15 static Win32CORE
>     CYG17 cyg\-1\.7 paths\-utf8
>     0c612ce82 Fix building static extensions on cygwin\, \-UUSEIMPORTLIB
>     1bac5ecc1 Fix 64\-bit threading sv\.c&#8203;: S\_anonymise\_cv\_maybe
>     Cygwin&#8203;::sync\_winenv added
>
> \-\-\-
> @&#8203;INC for perl 5\.14\.2&#8203;:
>     /cygdrive/f/perl/local/lib/perl5/cygwin\-thread\-multi\-64int
>     /cygdrive/f/perl/local/lib/perl5
>     /usr/lib/perl5/site\_perl/5\.14/i686\-cygwin\-threads\-64int
>     /usr/lib/perl5/site\_perl/5\.14
>     /usr/lib/perl5/vendor\_perl/5\.14/i686\-cygwin\-threads\-64int
>     /usr/lib/perl5/vendor\_perl/5\.14
>     /usr/lib/perl5/5\.14/i686\-cygwin\-threads\-64int
>     /usr/lib/perl5/5\.14
>     /usr/lib/perl5/site\_perl/5\.10
>     /usr/lib/perl5/vendor\_perl/5\.10
>     /usr/lib/perl5/site\_perl/5\.8
>     \.
>
> \-\-\-
> Environment for perl 5\.14\.2&#8203;:
>     HOME=/home/chm
>     LANG=C
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>
> PATH=/cygdrive/f/perl/local/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/cygdrive/c/Windows/system32
>     PERL5LIB=/cygdrive/f/perl/local/lib/perl5
>     PERLDOC\_PAGER=less
>     PERL\_BADLANG \(unset\)
>     PERL\_RL=perl
>     SHELL=/bin/bash
>
>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @kevindawson

now works as expected \( perl 5\.18\.1 \)

$ perl \-d 116358\.pl 

Loading DB routines from perl5db\.pl version 1\.39\_10
Editor support available\.

Enter h or 'h h' for help\, or 'man perldebug' for more help\.

main&#8203;::\(116358\.pl&#8203;:6\)&#8203;:    my $data = '';
&nbsp;  DB\<1> s
main&#8203;::\(116358\.pl&#8203;:15\)&#8203;:    print "Start test\, \\$data=$data\\n";
&nbsp;  DB\<1> 
Start test\, $data=
main&#8203;::\(116358\.pl&#8203;:16\)&#8203;:    lslice\(3\, 5\) = 4;
&nbsp;  DB\<1> 
main&#8203;::lslice\(116358\.pl&#8203;:9\)&#8203;:        my \($arg1\, $val1\) = @&#8203;\_;
&nbsp;  DB\<1> 
main&#8203;::lslice\(116358\.pl&#8203;:10\)&#8203;:        print "\.\. in lslice now\\n";
&nbsp;  DB\<1> 
\.\. in lslice now
main&#8203;::lslice\(116358\.pl&#8203;:11\)&#8203;:        print "\.\. $arg1 \+ $val1 = "\, $arg1 \+ $val1\,
"\\n";
&nbsp;  DB\<1> 
\.\. 3 \+ 5 = 8
main&#8203;::lslice\(116358\.pl&#8203;:12\)&#8203;:        $data;
&nbsp;  DB\<1> 
main&#8203;::\(116358\.pl&#8203;:17\)&#8203;:    print "Stop test\, \\$data=$data\\n";
&nbsp;  DB\<1> 
Stop test\, $data=4
main&#8203;::\(116358\.pl&#8203;:19\)&#8203;:    exit\(0\);
&nbsp;  DB\<1> 
Debugged program terminated\.  Use q to quit or R to restart\,
use o inhibit\_exit to avoid stopping after program termination\,
h q\, h R or h o to get additional info\.
&nbsp;  DB\<1> q

imho this is was fixed in shlom\-fish wake and thus closed

**Comment by p5pRT** [comment]
### From @kevindawson

[116358.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/1247903/651915/116358.pl)

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

now works as expected \( perl 5\.18\.1 \)

$ perl \-d 116358\.pl 

Loading DB routines from perl5db\.pl version 1\.39\_10
Editor support available\.

Enter h or 'h h' for help\, or 'man perldebug' for more help\.

main&#8203;::\(116358\.pl&#8203;:6\)&#8203;:    my $data = '';
&nbsp;  DB\<1> s
main&#8203;::\(116358\.pl&#8203;:15\)&#8203;:    print "Start test\, \\$data=$data\\n";
&nbsp;  DB\<1> 
Start test\, $data=
main&#8203;::\(116358\.pl&#8203;:16\)&#8203;:    lslice\(3\, 5\) = 4;
&nbsp;  DB\<1> 
main&#8203;::lslice\(116358\.pl&#8203;:9\)&#8203;:        my \($arg1\, $val1\) = @&#8203;\_;
&nbsp;  DB\<1> 
main&#8203;::lslice\(116358\.pl&#8203;:10\)&#8203;:        print "\.\. in lslice now\\n";
&nbsp;  DB\<1> 
\.\. in lslice now
main&#8203;::lslice\(116358\.pl&#8203;:11\)&#8203;:        print "\.\. $arg1 \+ $val1 = "\, $arg1 \+ $val1\,
"\\n";
&nbsp;  DB\<1> 
\.\. 3 \+ 5 = 8
main&#8203;::lslice\(116358\.pl&#8203;:12\)&#8203;:        $data;
&nbsp;  DB\<1> 
main&#8203;::\(116358\.pl&#8203;:17\)&#8203;:    print "Stop test\, \\$data=$data\\n";
&nbsp;  DB\<1> 
Stop test\, $data=4
main&#8203;::\(116358\.pl&#8203;:19\)&#8203;:    exit\(0\);
&nbsp;  DB\<1> 
Debugged program terminated\.  Use q to quit or R to restart\,
use o inhibit\_exit to avoid stopping after program termination\,
h q\, h R or h o to get additional info\.
&nbsp;  DB\<1> q

imho this is was fixed in shlom\-fish wake and thus closed

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Aug 26 09&#8203;:39&#8203;:30 2013\, bowtie wrote&#8203;:
> now works as expected \( perl 5\.18\.1 \)
> 
> $ perl \-d 116358\.pl 
> 
> Loading DB routines from perl5db\.pl version 1\.39\_10
> Editor support available\.
> 
> Enter h or 'h h' for help\, or 'man perldebug' for more help\.
> 
> main&#8203;::\(116358\.pl&#8203;:6\)&#8203;:    my $data = '';
>   DB\<1> s
> main&#8203;::\(116358\.pl&#8203;:15\)&#8203;:    print "Start test\, \\$data=$data\\n";
>   DB\<1> 
> Start test\, $data=
> main&#8203;::\(116358\.pl&#8203;:16\)&#8203;:    lslice\(3\, 5\) = 4;
>   DB\<1> 
> main&#8203;::lslice\(116358\.pl&#8203;:9\)&#8203;:        my \($arg1\, $val1\) = @&#8203;\_;
>   DB\<1> 
> main&#8203;::lslice\(116358\.pl&#8203;:10\)&#8203;:        print "\.\. in lslice now\\n";
>   DB\<1> 
> \.\. in lslice now
> main&#8203;::lslice\(116358\.pl&#8203;:11\)&#8203;:        print "\.\. $arg1 \+ $val1 = "\, $arg1 \+ $val1\,
> "\\n";
>   DB\<1> 
> \.\. 3 \+ 5 = 8
> main&#8203;::lslice\(116358\.pl&#8203;:12\)&#8203;:        $data;
>   DB\<1> 
> main&#8203;::\(116358\.pl&#8203;:17\)&#8203;:    print "Stop test\, \\$data=$data\\n";
>   DB\<1> 
> Stop test\, $data=4
> main&#8203;::\(116358\.pl&#8203;:19\)&#8203;:    exit\(0\);
>   DB\<1> 
> Debugged program terminated\.  Use q to quit or R to restart\,
> use o inhibit\_exit to avoid stopping after program termination\,
> h q\, h R or h o to get additional info\.
>   DB\<1> q
> 
> imho this is was fixed in shlom\-fish wake and thus closed

Actually\, this is the same as \#118839\, which was fixed in bf2614180a\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
